core/server/master/src/main/java/alluxio/master/FaultTolerantAlluxioMasterProcess.java

```diff
@@ -70,50 +69,43 @@ public void start() throws Exception {
     }
 
     while (!Thread.interrupted()) {
-      if (mServingThread == null) {
-        // We are in secondary mode. Nothing to do until we become the primary.
-        mLeaderSelector.waitForState(State.PRIMARY);
-        LOG.info("Transitioning from secondary to primary");
-        AbstractJournalSystem.ALLOW_JOURNAL_MODIFY.set(true);
-        mJournalSystem.setMode(Mode.PRIMARY);
-        stopMasters();
-        LOG.info("Secondary stopped");
-        startMasters(true);
-        mServingThread = new Thread(() -> {
-          try {
-            startServing(" (gained leadership)", " (lost leadership)");
-          } catch (Throwable t) {
+      startMasters(false);
+      LOG.info("Secondary started");
+      mLeaderSelector.waitForState(State.PRIMARY);
+      mJournalSystem.setMode(Mode.PRIMARY);
+      stopMasters();
+      LOG.info("Secondary stopped");
+
+      startMasters(true);
+      mServingThread = new Thread(() -> {
+        try {
+          startServing(" (gained leadership)", " (lost leadership)");
+        } catch (Throwable t) {
             Throwable root = ExceptionUtils.getRootCause(t);
-            if ((root != null && (root instanceof InterruptedException)) || Thread.interrupted()) {
-              return;
-            }
-            LOG.error("Exception thrown in main serving thread. System exiting.", t);
-            System.exit(-1);
+          if ((root != null && (root instanceof InterruptedException)) || Thread.interrupted()) {
+            return;
           }
-        }, "MasterServingThread");
-        mServingThread.start();
-        LOG.info("Primary started");
-      } else {
-        // We are in primary mode. Nothing to do until we become the secondary.
-        mLeaderSelector.waitForState(State.SECONDARY);
-        LOG.info("Transitioning from primary to secondary");
-        AbstractJournalSystem.ALLOW_JOURNAL_MODIFY.set(false);
-        stopServing();
-        mServingThread.join(mServingThreadTimeoutMs);
-        if (mServingThread.isAlive()) {
-          LOG.error(
-              "Failed to stop serving thread after {}ms. Printing serving thread stack trace "
-                  + "and exiting.\n{}",
-              mServingThreadTimeoutMs, ThreadUtils.formatStackTrace(mServingThread));
+          LOG.error("Exception thrown in main serving thread. System exiting.", t);
           System.exit(-1);
         }
-        mServingThread = null;
-        stopMasters();
-        mJournalSystem.setMode(Mode.SECONDARY);
-        LOG.info("Primary stopped");
-        startMasters(false);
-        LOG.info("Secondary started");
+      }, "MasterServingThread");
+      mServingThread.start();
+      waitForReady();
+      LOG.info("Primary started");
+      mLeaderSelector.waitForState(State.SECONDARY);
+      mJournalSystem.setMode(Mode.SECONDARY);
```

Comment:
This PR moves the journal fencing much sooner - previously we wouldn't set the journal to secondary until the master had stopped serving, which can take quite some time.

Reply:
can you add this to the comments in the code, about this ordering?

Reply:
done